File: mvp/sdk/typescript/src/api/presentation-gen.ts
import { BaseAPI } from "./base";
import {
  Presentation,
  PresentationCreateRequest,
  PresentationFromTranscriptRequest,
  PresentationExportRequest,
  PresentationSlide,
  PresentationTemplate,
  PaginationParams,
} from "../types";

/**
 * Presentation Gen API — AI slide generation with 5 templates,
 * export to HTML/MD/PDF.
 */
export class PresentationGenAPI extends BaseAPI {
  // -- CRUD ----------------------------------------------------------------

  /** Generate a presentation from a topic. */
  async create(data: PresentationCreateRequest): Promise<Presentation> {
    return this._post<Presentation>("/api/presentations", data);
  }

  /** List presentations. */
  async list(params?: PaginationParams): Promise<Presentation[]> {
    return this._get<Presentation[]>("/api/presentations", {
      limit: params?.limit,
      offset: params?.offset,
    });
  }

  /** Get presentation with slides. */
  async get(id: string): Promise<Presentation> {
    return this._get<Presentation>(`/api/presentations/${id}`);
  }

  /** Delete a presentation. */
  async delete(id: string): Promise<void> {
    return this._delete(`/api/presentations/${id}`);
  }

  // -- Slide editing -------------------------------------------------------

  /** Update a slide at a given position. */
  async updateSlide(presentationId: string, position: number, data: Partial<PresentationSlide>): Promise<PresentationSlide> {
    return this._put<PresentationSlide>(`/api/presentations/${presentationId}/slides/${position}`, data);
  }

  /** Insert a slide after a given position. */
  async insertSlide(presentationId: string, afterPosition: number, data: Partial<PresentationSlide>): Promise<PresentationSlide> {
    return this._post<PresentationSlide>(`/api/presentations/${presentationId}/slides/${afterPosition}`, data);
  }

  /** Remove a slide. */
  async removeSlide(presentationId: string, position: number): Promise<void> {
    return this._delete(`/api/presentations/${presentationId}/slides/${position}`);
  }

  // -- Export --------------------------------------------------------------

  /** Export presentation (HTML, Markdown, or PDF). */
  async export(id: string, data: PresentationExportRequest): Promise<string> {
    return this._post<string>(`/api/presentations/${id}/export`, data);
  }

  // -- From transcript -----------------------------------------------------

  /** Generate a presentation from a transcription. */
  async fromTranscript(data: PresentationFromTranscriptRequest): Promise<Presentation> {
    return this._post<Presentation>("/api/presentations/from-transcript", data);
  }

  // -- Templates -----------------------------------------------------------

  /** List available presentation templates. */
  async listTemplates(): Promise<PresentationTemplate[]> {
    return this._get<PresentationTemplate[]>("/api/presentations/templates");
  }
}

File: mvp/sdk/python/saas_ia/api/presentation_gen.py
"""Presentation Gen API — AI slide generation (5 templates)."""

from __future__ import annotations

from typing import Any

from saas_ia.api.base import BaseAPI


class PresentationGenAPI(BaseAPI):
    """Wraps ``/api/presentations`` endpoints."""

    async def create(
        self,
        topic: str,
        *,
        template: str | None = None,
        slide_count: int | None = None,
        language: str | None = None,
    ) -> dict[str, Any]:
        """Generate a presentation from a topic."""
        body: dict[str, Any] = {"topic": topic}
        if template:
            body["template"] = template
        if slide_count:
            body["slide_count"] = slide_count
        if language:
            body["language"] = language
        return await self._post("/api/presentations", json=body)

    async def list(
        self,
        *,
        limit: int | None = None,
        offset: int | None = None,
    ) -> list[dict[str, Any]]:
        """List presentations."""
        return await self._get(
            "/api/presentations",
            params={"limit": limit, "offset": offset},
        )

    async def get(self, presentation_id: str) -> dict[str, Any]:
        """Get presentation with slides."""
        return await self._get(f"/api/presentations/{presentation_id}")

    async def delete(self, presentation_id: str) -> None:
        """Delete a presentation."""
        await self._delete(f"/api/presentations/{presentation_id}")

    async def update_slide(
        self,
        presentation_id: str,
        position: int,
        *,
        title: str | None = None,
        content: str | None = None,
        notes: str | None = None,
        layout: str | None = None,
    ) -> dict[str, Any]:
        """Update a slide at a given position."""
        body: dict[str, Any] = {}
        if title is not None:
            body["title"] = title
        if content is not None:
            body["content"] = content
        if notes is not None:
            body["notes"] = notes
        if layout is not None:
            body["layout"] = layout
        return await self._put(
            f"/api/presentations/{presentation_id}/slides/{position}",
            json=body,
        )

    async def insert_slide(
        self,
        presentation_id: str,
        after_position: int,
        **slide_data: Any,
    ) -> dict[str, Any]:
        """Insert a slide after a given position."""
        return await self._post(
            f"/api/presentations/{presentation_id}/slides/{after_position}",
            json=slide_data,
        )

    async def remove_slide(
        self, presentation_id: str, position: int
    ) -> None:
        """Remove a slide."""
        await self._delete(
            f"/api/presentations/{presentation_id}/slides/{position}"
        )

    async def export(
        self,
        presentation_id: str,
        format: str = "html",
    ) -> Any:
        """Export presentation (html, markdown, pdf)."""
        return await self._post(
            f"/api/presentations/{presentation_id}/export",
            json={"format": format},
        )

    async def from_transcript(
        self,
        transcription_id: str,
        *,
        template: str | None = None,
        slide_count: int | None = None,
    ) -> dict[str, Any]:
        """Generate a presentation from a transcription."""
        body: dict[str, Any] = {"transcription_id": transcription_id}
        if template:
            body["template"] = template
        if slide_count:
            body["slide_count"] = slide_count
        return await self._post("/api/presentations/from-transcript", json=body)

    async def list_templates(self) -> list[dict[str, Any]]:
        """List available presentation templates."""
        return await self._get("/api/presentations/templates")
